src/Data/List/Base.agda

```diff
@@ -195,6 +195,15 @@ tails : List A → List (List A)
 tails []       = [] ∷ []
 tails (x ∷ xs) = (x ∷ xs) ∷ tails xs
 
+insertAt : (xs : List A) → Fin (suc (length xs)) → A → List A
+insertAt []       zero    v = [ v ]
+insertAt (x ∷ xs) zero    v = v ∷ x ∷ xs
```

Comment:
The single clause `insertAt xs zero v = v ∷ xs` solves many of the problems you have been having.